js/private/js_library_helpers.bzl

```diff
@@ -27,39 +27,41 @@ def gather_transitive_sources(sources, targets):
     """Gathers transitive sources from a list of direct sources and targets
 
     Args:
-        sources: direct sources which should be included in `transitive_sources`
+        sources: list or depset of direct sources which should be included in `transitive_sources`
         targets: list of targets to gather `transitive_sources` from `JsInfo`
 
     Returns:
-        List of transitive sources
+        Depset of transitive sources
     """
-    transitive_sources = sources + [
-        item
+    transitive_deps = [
+        target[JsInfo].transitive_sources
         for target in targets
-        if JsInfo in target and hasattr(target[JsInfo], "transitive_sources")
-        for item in target[JsInfo].transitive_sources
+        if JsInfo in target
     ]
-    transitive_sources = [file for file in transitive_sources if file]
-    return transitive_sources
+    if type(sources) == "list":
+      return depset(sources, transitive = transitive_deps)
+    else:
+      return depset([], transitive = [sources] + transitive_deps)
 
 def gather_transitive_declarations(declarations, targets):
     """Gathers transitive sources from a list of direct sources and targets
 
     Args:
-        declarations: Direct sources which should be included in `transitive_declarations`
+        declarations: list or depset of direct sources which should be included in `transitive_declarations`
         targets: List of targets to gather `transitive_declarations` from `JsInfo`
 
     Returns:
-        List of transitive sources
+        Depset of transitive sources
     """
-    transitive_declarations = declarations + [
-        item
+    transitive_deps = [
+        target[JsInfo].transitive_declarations
         for target in targets
-        if JsInfo in target and hasattr(target[JsInfo], "transitive_declarations")
-        for item in target[JsInfo].transitive_declarations
+        if JsInfo in target
     ]
-    transitive_declarations = [file for file in transitive_declarations if file]
-    return transitive_declarations
+    if type(declarations) == "list":
+      return depset(declarations, transitive = transitive_deps)
```

Comment:
Should we do `depset([], transitive = [depset(declarations)] + transitive_deps)` so it is equivelent to the `else`?

I'm not sure if it really matters but have this slight difference could cause a difference downstream?

Reply:
That's definitely doable. I think the only difference would be in the ordering of the files when pulled out (one treats it as siblings instead of parents). While waiting on approving the CLA, I'll see if anyone has a good reason to do this (or not to do this).

Reply:
Made this change.